karabo_data/stacking.py

```diff
@@ -0,0 +1,208 @@
+        if len(shapes) > 2 or (s1[0] != 0) or (s1[1:] != s2[1:]):
+            raise ValueError("Arrays have mismatched shapes: {}".format(shapes))
+        empty_mods = {n for n, a in modno_arrays.items() if a.shape == s1}
+        for modno in empty_mods:
+            del modno_arrays[modno]
+        shapes.remove(s1)
+    if max(modno_arrays) >= modules:
+        raise IndexError("Module {} is out of range for a detector with {} modules"
+                         .format(max(modno_arrays), modules))
+
+    dtype = dtypes.pop()
+    shape = shapes.pop()
+    stack = VirtualStack(
+        modno_arrays, modules, shape, dtype, fillvalue, stack_axis=axis
+    )
+    if virtual:
+        return stack
+
+    return stack.asarray()
+
+
+class VirtualStack:
+    """'Virtual' array for stacking detector data from several modules.
+
+    Access is limited to either a single module at a time or all modules
+    together, but this is enough to assemble detector images.
+    """
+    def __init__(self, data, nmodules, mod_shape, dtype, fillvalue,
+                 stack_axis=-3):
+        self._nmodules = nmodules
+        self._data = data  # {modno: array}
+        self.dtype = dtype
+        self._fillvalue = fillvalue
+        self._mod_shape = mod_shape
+        self.ndim = len(mod_shape) + 1
+        self._stack_axis = stack_axis
+        if self._stack_axis < 0:
+            self._stack_axis += self.ndim
+        sax = self._stack_axis
+        self.shape = mod_shape[:sax] + (nmodules,) + mod_shape[sax:]
+
+    def __repr__(self):
+        return "<VirtualStack (shape={}, {}/{} modules, dtype={})>".format(
+            self.shape, len(self._data), self._nmodules, self.dtype,
+        )
+
+    # Multidimensional slicing
+    def __getitem__(self, slices):
+        if not isinstance(slices, tuple):
+            slices = (slices,)
+
+        missing_dims = self.ndim - len(slices)
+        if Ellipsis in slices:
+            ix = slices.index(Ellipsis)
+            missing_dims += 1
+            slices = slices[:ix] + (slice(None, None),) * missing_dims + slices[ix + 1:]
+        else:
+            slices = slices + (slice(None, None),) * missing_dims
+
+        modno = slices[self._stack_axis]
+        mod_slices = slices[:self._stack_axis] + slices[self._stack_axis + 1:]
+
+        if isinstance(modno, int):
+            if modno < 0:
+                modno += self._nmodules
+            return self._get_single_mod(modno, mod_slices)
+        elif modno == slice(None, None):
+            return self._get_all_mods(mod_slices)
+        else:
+            raise Exception(
+                "VirtualStack can only slice a single module or all modules"
+            )
+
+    def _get_single_mod(self, modno, mod_slices):
+        try:
+            mod_data = self._data[modno]
+        except KeyError:
+            if modno >= self._nmodules:
+                raise IndexError(modno)
+            mod_data = np.full(self._mod_shape, self._fillvalue, self.dtype)
```

Comment:
Does it make sense also to set 
```
self._data[modno] = mod_data
```
although I don't know the use case of getting a missing module twice.

here?

Reply:
I'm not sure the use case either. It would make the behaviour more consistent if you do something like `stack[0][100:200, 100:200] = 2.0`, though. At present, the effect of that is different if module 0 is there or not.

Reply:
This is a good reason. Thanks!